<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 <HEAD>
   <TITLE> 1.1.1b crash (RUN_ONCE problem?)
   </TITLE>
   <LINK REL="Index" HREF="https://mta.openssl.org/pipermail/openssl-users/2019-March/index.html" >
   <LINK REL="made" HREF="mailto:openssl-users%40openssl.org?Subject=Re%3A%201.1.1b%20crash%20%28RUN_ONCE%20problem%3F%29&In-Reply-To=%3C40BB7E4A-A183-4759-BEEB-5510F6765948%40dukhovni.org%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="010036.html">
   <LINK REL="Next"  HREF="010006.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>1.1.1b crash (RUN_ONCE problem?)</H1>
    <B>Viktor Dukhovni</B> 
    <A HREF="mailto:openssl-users%40openssl.org?Subject=Re%3A%201.1.1b%20crash%20%28RUN_ONCE%20problem%3F%29&In-Reply-To=%3C40BB7E4A-A183-4759-BEEB-5510F6765948%40dukhovni.org%3E"
       TITLE="1.1.1b crash (RUN_ONCE problem?)">openssl-users at dukhovni.org
       </A><BR>
    <I>Tue Mar  5 04:41:36 UTC 2019</I>
    <P><UL>
        <LI>Previous message: <A HREF="010036.html">1.1.1b crash (RUN_ONCE problem?)
</A></li>
        <LI>Next message: <A HREF="010006.html">OpenSSL 3.0 vs. SSL 3.0
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#10038">[ date ]</a>
              <a href="thread.html#10038">[ thread ]</a>
              <a href="subject.html#10038">[ subject ]</a>
              <a href="author.html#10038">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>&gt;<i> On Mar 4, 2019, at 11:29 PM, Norm Green &lt;<A HREF="../../../mailman/listinfo/openssl-users.html">norm.green at gemtalksystems.com</A>&gt; wrote:
</I>&gt;<i> 
</I>&gt;<i> Yes I'm sure the build process is correct.
</I>&gt;<i> Turns out this problem was cause by one thread calling exit() while another thread was doing SSL_write().  The SSL exit handler triggered by exit() was causing the lock in question to be freed AFAIKT.
</I>&gt;<i> So it would seem that threads either need to exit with pthread_exit() or return to a known state such that no SSL calls are in progress in any thread before the process can safely exit().
</I>
Yes, main() must not exit while threads are running code from
libraries with exit handlers.  This is issue is not unique to
OpenSSL.  I've observed the same problem with threading running
code in MIT's GSSAPI library.

Some users want libraries to be unloadable without memory leaks,
others want safe exit() while threads are running...  Serving
both is hard.

A multi-threaded program that wants to exit before all the
threads are done, should probably call _exit(), and skip
the exit handlers (when none of those are doing anything
critical, like flushing stdio buffers).

Alternatively arrange for a way to ask threads to terminate
promptly, and wait for them to do that.

-- 
-- 
	Viktor.

</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="010036.html">1.1.1b crash (RUN_ONCE problem?)
</A></li>
	<LI>Next message: <A HREF="010006.html">OpenSSL 3.0 vs. SSL 3.0
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#10038">[ date ]</a>
              <a href="thread.html#10038">[ thread ]</a>
              <a href="subject.html#10038">[ subject ]</a>
              <a href="author.html#10038">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="../../../mailman/listinfo/openssl-users.html">More information about the openssl-users
mailing list</a><br>
</body></html>
